docs/buffers.md

```diff
@@ -0,0 +1,137 @@
+1. Telling the kernel that one of its arguments is that specific buffer
+2. Set correct buffer attributes (eg. global memory id) according to the kernel argument's attribute.
+3. Create and bind the host channel if the kernel arg is a host pipe.
+
+Important: Given SYCL calls `clSetKernelArg` after `clEnqueueWriteBuffer`. Make sure to never change buffer properties inside `clSetKernelArg`.
+
+### Enqueue Kernel / Task
+
+During [kernel enqueue](https://github.com/intel/fpga-runtime-for-opencl/blob/3f7a228133f92c63be5b04e222f3fc8ff72310e6/src/acl_kernel.cpp#L1644-L2313), it will call [l_copy_and_adjust_arguments_for_device](https://github.com/intel/fpga-runtime-for-opencl/blob/3f7a228133f92c63be5b04e222f3fc8ff72310e6/src/acl_kernel.cpp#L2730-L2983) to:
+1. Create a temporary buffer object that is the aligned copy of buffer arg value
+2. Get the correct kernel's required buffer location
+3. Reserve space at the required device global memory if not already reserved
+4. Copy the reserved address into the kernel invocation image
+5. Prepare memory migration information containing this temporary buffer as the source, destination device id as the target. Note at this point, the temporary buffer could have already gone through memory transfer. This is being taken care of when actually migrating the buffer ([`acl_mem_migrate_buffer`](https://github.com/intel/fpga-runtime-for-opencl/blob/b08e0af97351718ce0368a9ee507242b35f4929e/src/acl_mem.cpp#L5412-L5665)), i.e the function will know whether it should be moving from host to device or device to device (generally device to device operation is faster).
+
+
+Some other note about memory operation during enqueue:
+1. Device local pointer size is 4, device global pointer size is always the device's address bit integer division 8
+
+Before [submitting kernel](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L2982-L3093) to device queue. It first checks if the device is programmed, if not, it will [queue reprogram device operation](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3034) to do so. Then it will [arrange memory migration](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3043) for each kernel memory argument.
+
+Different from the device operation resulting from enqueue read/write, the memory migration calls on [`acl_mem_migrate_buffer`](https://github.com/intel/fpga-runtime-for-opencl/blob/b08e0af97351718ce0368a9ee507242b35f4929e/src/acl_mem.cpp#L5412-L5665) (i.e memory transfer and memory migration behave differently). 
+
+In memory migration:
```

Comment:
shouldn't this also be a section ?

Reply:
Thanks for the approval! I have wrapped memory migration into a subsection.